### PR TITLE
DP-107 Use messaging abstractions for AWS integration (WIP)

### PR DESCRIPTION
**Pull request by clicktravel-basharat**
Signed-off-by: Steffan Westcott <steffan.westcott@clicktravel.com>

### CODE REVIEW

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/DefaultSnsTopicFactory.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            topicArn = createAwsSnsTopic(name);
+            final SnsTopic snsTopic = new SnsTopic(name, topicArn, amazonSnsClient);
+            snsTopic.setPolicy(defaultPolicy(snsTopic)); // policy is set once, on creation of AWS SNS topic
+            return snsTopic;
+        } else {
+            return new SnsTopic(name, topicArn, amazonSnsClient);
+        }
+    }
+
+    // TODO : Give this method a better name. What does this policy do?
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/DefaultSnsTopicFactory.java]: This policy allows any queue to "bind" to this topic.

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopic.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import com.amazonaws.services.sns.model.PublishRequest;
+import com.amazonaws.services.sns.model.SetTopicAttributesRequest;
+import com.amazonaws.services.sns.model.SubscribeRequest;
+import com.clicktravel.cheddar.infrastructure.messaging.exception.MessagePublishException;
+
+/**
+ * A representative for an actual existing AWS SNS topic. Provides some convenience methods for working with the actual
+ * AWS SNS topic.
+ */
+public class SnsTopic {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopic.java]: SNS Topic should implement Exchange

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueue.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import com.amazonaws.auth.policy.Policy;
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.model.*;
+import com.clicktravel.cheddar.infrastructure.messaging.exception.MessageSendException;
+
+/**
+ * A representative for an actual existing AWS SQS queue. Provides some convenience methods for working with the actual
+ * AWS SQS queue.
+ */
+public class SqsQueue {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueue.java]: Shouldn't SqsQueue implement MessageQueue ?

**Comment by clicktravel-basharat** [comment]
First pass done. I think there might be too much abstraction between a SNS Topic (which is the AWS implementation of a messaging exchange) and "Exchange" and also between SQS Queue and "Queue" via intermediatary objects SnsExchange and SqsMessageQueue

**Comment by clicktravel-basharat** [comment]
Rename SqsQueue -> SqsResource
Rename SnsTopic -> SnsResource

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopic.java (hunk 2 above)
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopic.java]: No, the class SnsExchange<T> implements Exchange<T>.
SnsTopic is a supporting class which represents the underlying AWS resource used to implement SnsExchange<T>.

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueue.java (hunk 3 above)
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueue.java]: Similar to the above, `SqsMessageQueue<T>` implements `MessageQueue<T>`.
`SqsQueue` is a supporting class which represents the underlying AWS resource used to implement `SqsMessageQueue<T>`.

**Comment by clicktravel-steffan** [comment]
The suggested renames may be a little vague. How about instead:
`SqsQueue` becomes `SqsQueueResource`
`SnsTopic` becomes `SnsTopicResource`

**Comment by clicktravel-basharat** [comment]
OK with suggested names.

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat I've renamed the classes as discussed. I've also fixed a bug found on recent inspection. Ready for second review.

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/DefaultSnsTopicFactory.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import com.amazonaws.auth.policy.*;
+import com.amazonaws.auth.policy.Statement.Effect;
+import com.amazonaws.auth.policy.actions.SNSActions;
+import com.amazonaws.services.sns.AmazonSNS;
+import com.amazonaws.services.sns.model.CreateTopicRequest;
+import com.amazonaws.services.sns.model.ListTopicsResult;
+import com.amazonaws.services.sns.model.Topic;
+
+@Component
+public class DefaultSnsTopicFactory implements SnsTopicFactory {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/DefaultSnsTopicFactory.java]: DefaultSnsTopicFactory -> DefaultSnsTopicResourceFactory
SnsTopicFactory -> SnsTopicResourceFactory

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/DefaultSqsQueueFactory.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import com.amazonaws.auth.policy.conditions.ConditionFactory;
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.model.CreateQueueRequest;
+import com.amazonaws.services.sqs.model.GetQueueUrlRequest;
+import com.amazonaws.services.sqs.model.QueueAttributeName;
+import com.amazonaws.services.sqs.model.QueueDoesNotExistException;
+import com.clicktravel.common.functional.StringUtils;
+
+@Component
+public class DefaultSqsQueueFactory implements SqsQueueFactory {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/DefaultSqsQueueFactory.java]: DefaultSqsQueueFactory -> DefaultSqsQueueResourceFactory
SqsQueueFactory -> SqsQueueResourceFactory

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopicFactory.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.infrastructure.messaging.aws;
+
+public interface SnsTopicFactory {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopicFactory.java]: SnsTopicFactory -> SnsTopicResourceFactory

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueueFactory.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.infrastructure.messaging.aws;
+
+public interface SqsQueueFactory {
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueueFactory.java]: SqsQueueFactory -> SqsResourceQueueFactory

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueueFactory.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    /**
+     * Creates an {@link SqsQueueResource} which is subscribed to the given {@link SnsTopicResource}s. An actual AWS SQS queue is
+     * created if one does not already exist with the given name. The topic subscriptions (and a compatible queue
+     * policy) are always updated, whether or not the AWS SQS queue already exists.
+     * @param name Queue name
+     * @param snsTopics {@link SnsTopicResource}s to subscribe to
+     * @return {@link SqsQueueResource} with the given name, subscribed to the specified topics.
+     */
+    SqsQueueResource createSqsQueue(String name, SnsTopicResource... snsTopics);
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SqsQueueFactory.java]: createSqsQueue -> createSqsQueueResource

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopicFactory.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Creates a {@link SnsTopicResource} and the actual AWS SNS topic if it does not already exist with the given name.
+     * @param name Topic name
+     * @return {@link SnsTopicResource} with the given name
+     */
+    SnsTopicResource createSnsTopicAndAwsSnsTopicIfAbsent(String name);
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopicFactory.java]: createSnsTopicAndAwsSnsTopicIfAbsent -> createSnsTopicResourceAndAwsSnsTopicIfAbsent

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopicFactory.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Creates a {@link SnsTopicResource} representing an existing actual AWS SNS topic.
+     * @param name Topic name
+     * @return {@link SnsTopicResource} with the given name
+     */
+    SnsTopicResource createSnsTopicForExistingAwsSnsTopic(String name);
```
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/messaging/aws/SnsTopicFactory.java]: createSnsTopicResourceForExistingAwsSnsTopic

**Comment by clicktravel-basharat** [comment]
More naming changes.

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat Renaming done, ready for review.

**Comment by clicktravel-basharat** [comment]
All good.

**Pull request merged**